### PR TITLE
doc: extensions: improve doxyrunner

### PR DESCRIPTION
**Pull request by gmarull**
```
doc: extensions: doxyrunner: use file hashes for cache

Use file hashes instead of paths and modification times. This method
fits better on systems using cache mechanisms. Note that hash is
computed using content obtained in UTF-8 text mode for portability
reasons.
```
```
doc: extensions: doxyrunner: store cache in Sphinx environment

Sphinx provides a way to persist data across builds: the
BuildEnvironment. The build environment is automatically managed by
Sphinx, so there is no need to take care of loading/dumping cache files.
```

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]
LGTM

File: doc/_extensions/zephyr/doxyrunner.py [hunk 1]
```diff
@@ -169,16 +185,16 @@ def doxygen_input_has_changed(doxyfile: str, cache_dir: Path) -> bool:
...
 
-    # build a dict of input files <-> current modification time
-    files = dict()
+    # build a set with input files hash
+    files = set()
     for file in input_files:
         path = Path(file)
         if path.is_file():
-            files[path.as_posix()] = path.stat().st_mtime_ns
+            files.add(hash_file(path))
```
Comment (mbolivar-nordic) [inline doc/_extensions/zephyr/doxyrunner.py]: What is the performance impact?

Comment (gmarull) [reply doc/_extensions/zephyr/doxyrunner.py]: I've just done a quick measurement:

Previous method: 0.032463788986206055s
New method: 0.07107734680175781 s

Comment (mbolivar-nordic) [reply doc/_extensions/zephyr/doxyrunner.py]: 2x slowdown!? We can't possibly accept that :stuck_out_tongue_winking_eye:

Just kidding. Thanks for measuring.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by gmarull** [comment]
rebased to see if CI gets _unblocked_

**Pull request merged**